src/DynamoCoreWpf/ViewModels/Watch3D/shaderSource/psDynamoMesh.hlsl

```diff
@@ -0,0 +1,127 @@
+
+#define MESH
+#include"Common.hlsl"
+#include"CommonBuffers.hlsl"
+#include"DynamoCommonStructures.hlsl"
+
+//--------------------------------------------------------------------------------------
+// Phong Lighting Reflection Model
+//--------------------------------------------------------------------------------------
+// Returns the sum of the diffuse and specular terms in the Phong reflection model
+// The specular and diffuse reflection constants for the currently loaded material (k_d and k_s)
+float4 calcPhongLighting(float4 LColor, float4 vMaterialTexture, float3 N, float3 L, float3 V, float3 R)
+{
+	float4 Id = vMaterialTexture * float4(1, 1, 1, 1) * saturate(dot(N, L));
+	float4 Is = vMaterialSpecular * pow(saturate(dot(R, V)), sMaterialShininess);
+	return (Id + Is) * LColor;
+}
+
+//used for texture lookup if the material has a diffuse map
+//for our use case this is geometry generated by Display.BySurfaceColors()
+SamplerState LinearSampler
+{
+	Filter = MIN_MAG_MIP_LINEAR;
+	AddressU = Wrap;
+	AddressV = Wrap;
+	MaxAnisotropy = 16;
+};
+
+float4 main(PSInputCustom input) : SV_Target
+{
+	//TODO move to common or pass in from material or the vParams.
+	float4 vSelectionColor = float4(0.0, 0.62, 1.0, 1.0);
+
+
+	//flags were passed through from vertex shader
+	//lets decode them
+	uint flags = int(input.customParams.x);
+
+	//our flags are packed in this order:
+  /*
+	  None = 0,
+	  IsFrozen = 1,
+	  IsSelected = 2,
+	  IsIsolated = 4,
+	  IsSpecialRenderPackage = 8, //not used
+	  HasTransparency = 16, //not used
+	  RequiresPerVertexColoration = 32
+	  FlatShade = 64 //not used
+	  */
+
+	bool isFrozen = flags & 1;
+	bool isSelected = flags & 2;
+	bool isIsolated = flags & 4;
+	bool isSpecialRenderPackage = flags & 8;
+	bool requiresPerVertexColoration = flags & 32;
+
+
+	//renormalize - does not support normal maps
+	input.n = normalize(input.n);
+
+	// get per pixel vector to eye-position
+	float3 eye = normalize(vEyePos - input.wp.xyz);
+
+	//we need to support diffuse masps for nodes like Display.BySurfaceColors which use diffuse maps
+	//instead of vertex colors.
+	if (bHasDiffuseMap)
+	{
+		float4 vMaterialTexture = 1.0f;
+		float4 I = vMaterialTexture = texDiffuseMap.Sample(LinearSampler, input.t);
+		//TODO.... do want to lerp or just return selection color.
+		if (isSelected) {
+			I = lerp(vSelectionColor, I, 0.3);
+		}
+
+		//TODO this is why lighting is not applied to textured geometry - we just return here...
+		//TODO use flat shade flag instead so users can choose to apply shading or not to any Display object.
+		return I;
+	}
+
+	// light emissive and ambient intensity
+	// this variable can be used for light accumulation
+	float4 I = vMaterialEmissive + vMaterialAmbient * vLightAmbient;
```

Comment:
Where are these defined - `vMaterialEmissive` etc.?

Reply:
these are defined in the common buffer for mesh and are set by the material that the mesh currently has assigned:
https://github.com/helix-toolkit/helix-toolkit/blob/76dc523c9bf928ef2c6f6e72246df28959cc0336/Source/Examples/WPF.SharpDX/CustomShaderDemo/Shaders/CommonBuffers.hlsl#L69